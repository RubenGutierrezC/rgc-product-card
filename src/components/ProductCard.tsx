import React, { createContext, CSSProperties } from 'react';
import styles from '../styles/styles.module.css';
import useProduct from '../hooks/useProduct';
import { InitialValues, ProductCartHandlers } from '../interface/interface';
import {
  ProductContextProps,
  Product,
  OnChangeArgs,
} from '../interface/interface';

export const ProductContext = createContext({} as ProductContextProps);
const { Provider } = ProductContext;

export interface Props {
  product: Product;
  className?: string;
  style?: CSSProperties;
  onChange?: (arg: OnChangeArgs) => void;
  value?: number;
  initialValues?: InitialValues;
  children?: (args: ProductCartHandlers) => JSX.Element;
}

export const ProductCard = ({
  children,
  product,
  className,
  style,
  onChange,
  value,
  initialValues,
}: Props) => {
  const {
    counter,
    increaseBy,
    maxCount,
    isMaxCountReached,
    reset,
  } = useProduct({
    onChange,
    product,
    value,
    initialValues,
  });

  return (
    <Provider value={{ product, increaseBy, counter, maxCount }}>
      <div className={`${styles.productCard} ${className}`} style={style}>
        {children &&
          children({
            count: counter,
            maxCount,
            isMaxCountReached,
            reset,
            product,
            increaseBy,
          })}
      </div>
    </Provider>
  );
};
