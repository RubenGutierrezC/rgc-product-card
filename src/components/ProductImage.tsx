import React, { useContext, CSSProperties } from 'react';
import { ProductContext } from './ProductCard';
import noImage from '../assets/no-image.jpg';
import styles from '../styles/styles.module.css';

export interface ProductImageProps {
  image?: string;
  className?: string;
  style?: CSSProperties;
}

export const ProductImage = ({
  image,
  className,
  style,
}: ProductImageProps) => {
  const { product } = useContext(ProductContext);

  let imgToShow: string;

  if (image) {
    imgToShow = image;
  } else if (product?.image) {
    imgToShow = product?.image;
  } else {
    imgToShow = noImage;
  }

  return (
    <img
      className={`${styles.productImg} ${className}`}
      src={imgToShow}
      alt="Product"
      style={style}
    />
  );
};
