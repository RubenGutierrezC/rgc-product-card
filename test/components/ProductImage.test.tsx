import React from 'react';
import renderer from 'react-test-renderer';
import { ProductCard, ProductImage } from '../../src/components';
import { product2 } from '../data/products';

describe('ProductImage', () => {
  test('debe mostrar el componente correctamente con la imagen personalizada', () => {
    const wrapper = renderer
      .create(
        <ProductImage
          image="https://picsum.photos/200"
          className="custom-class"
        />
      )
      .toJSON();

    expect(wrapper).toMatchSnapshot();
  });

  test('debe de mostrar el component con la imagen del product', () => {
    const wrapper = renderer.create(
      <ProductCard product={product2}>{() => <ProductImage />}</ProductCard>
    );

    expect(wrapper.toJSON()).toMatchSnapshot();
  });
});
