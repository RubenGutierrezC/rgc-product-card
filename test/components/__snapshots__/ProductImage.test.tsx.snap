// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductImage debe de mostrar el component con la imagen del product 1`] = `
<div
  className="productCard undefined"
>
  <img
    alt="Product"
    className="productImg undefined"
    src="https://picsum.photos/200"
  />
</div>
`;

exports[`ProductImage debe mostrar el componente correctamente con la imagen personalizada 1`] = `
<img
  alt="Product"
  className="productImg custom-class"
  src="https://picsum.photos/200"
/>
`;
